proposals/utils/pdf_diff_sections.py

```diff
@@ -1067,39 +786,106 @@ def create_context_pdf(context, proposal):
+    Many of these class methods receive a proposal explicitly and not from self,
+    so that they can be re-used for creating sections for the diff
+    """
 
-def multi_sections(section_type, objects, num=None):
-    """A function that creates a list of sections of a specified type, from a list of
-    objects, where it is possible for an object, such as a study, to be missing."""
-
-    if num is None:
-        return [section_type(obj) if obj is not None else None for obj in objects]
-    else:
-        return [section_type(obj, num) if obj is not None else None for obj in objects]
-
-
-def get_all_related(objects, related_name):
-    """Helper function for retrieving related objects"""
-
-    return [
-        getattr(obj, related_name, None) if obj is not None else None for obj in objects
-    ]
-
-
-def get_all_related_set(objects, related_name):
-    """Helper function for retrieving a set of related objects."""
-
-    return [
-        getattr(obj, related_name, None).all() if obj is not None else None
-        for obj in objects
-    ]
+    def __init__(self, proposal):
+        self.proposal = proposal
+
+    def get_all_attachment_sections(self):
+        """
+        Returns a list of Attachment sections, along with TitleSections to organize
+        the sections. Only used for PDF.
+        """
+        attachment_sections = []
+
+        slot_dict = self._all_attachments_dict(self.proposal)
+
+        for owner in slot_dict:
+            if slot_dict[owner]:
+                title = self._create_object_heading(owner, self.proposal)
+                attachment_sections.append(TitleSection(title))
+                for slot in slot_dict[owner]:
+                    attachment_sections.append(self._create_attachment_section(slot))
+
+        return attachment_sections
+
+    def _all_slots_list(self, proposal):
+        """
+        Returns a list of all slots for a specific proposal
+        """
+        from proposals.utils.stepper import Stepper
+
+        stepper = Stepper(proposal)
+        slots = [slot for slot in stepper.attachment_slots if slot.attachment]
+
+        return slots
+
+    def _all_owners_list(self, proposal):
+        """
+        Returns a list of all possible owners of attachments for a given proposal
+        """
+        return [proposal] + list(proposal.study_set.all())
+
+    def _all_attachments_dict(self, proposal):
+        """
+        Returns a dictionary where the objects are the keys and the attachments
+        are stored in a list as the value
+        """
+
+        slots = self._all_slots_list(proposal)
+
+        objects = self._all_owners_list(proposal)
+        slot_dict = {obj: [] for obj in objects}
+        for slot in slots:
+            relevant_owner = slot.attachment.get_owner_for_proposal(proposal)
+            slot_dict[relevant_owner].append(slot)
+
+        return slot_dict
+
+    def _create_object_heading(self, owner, proposal):
```

Comment:
I believe the "All Attachments" setting still needs a heading of its own. Perhaps just some kind of "TitleSection".